pandas/core/arrays/string_.py

```diff
@@ -0,0 +1,217 @@
+class StringDtype(ExtensionDtype):
+    """
+    Extension dtype for text data.
+
+    .. versionadded:: 1.0.0
+
+    Attributes
+    ----------
+    None
+
+    Methods
+    -------
+    None
+
+    Examples
+    --------
+    >>> pd.StringDtype()
+    StringDtype
+    """
+
+    @property
+    def na_value(self):
+        """
+        StringDtype uses :attr:`numpy.nan` as the missing NA value.
+        """
+        return np.nan
+
+    @property
+    def type(self) -> Type:
+        return str
+
+    @property
+    def name(self) -> str:
+        """
+        The alias for StringDtype is ``'string'``.
+        """
+        return "string"
+
+    @classmethod
+    def construct_from_string(cls, string: str):
+        if string == "string":
+            return cls()
+        return super().construct_from_string(string)
+
+    @classmethod
+    def construct_array_type(cls) -> "Type[StringArray]":
+        return StringArray
+
+    def __repr__(self) -> str:
+        return "StringDtype"
+
+
+class StringArray(PandasArray):
+    """
+    Extension array for text data.
+
+    .. versionadded:: 1.0.0
+
+    Parameters
+    ----------
+    values : array-like
+        The array of data.
+
+        .. warning::
+
+           Currently, this expects an object-dtype ndarray
+           where the elements are Python strings. This may
+           change without warning in the future.
+    copy : bool, default False
+        Whether to copy the array of data.
+
+    Attributes
+    ----------
+    None
+
+    Methods
+    -------
+    None
+
+    See Also
+    --------
+    Series.str
+        The string methods are available on Series backed by
+        a StringArray.
+
+    Examples
+    --------
+    >>> pd.array(['This is', 'some text', None, 'data.'], dtype="string")
+    <StringArray>
+    ['This is', 'some text', nan, 'data.']
+    Length: 4, dtype: string
+
+    Unlike ``object`` dtype arrays, ``StringArray`` doesn't allow non-string
+    values.
+
+    >>> pd.array(['1', 1], dtype="string")
+    Traceback (most recent call last):
+    ...
+    ValueError: Must provide strings
+    """
+
+    # undo the PandasArray hack
+    _typ = "extension"
+
+    def __init__(self, values, copy=False):
+        super().__init__(values, copy=copy)
+        self._dtype = StringDtype()
+        self._validate()
+
+    def _validate(self):
+        """Validate that we only store NA or strings."""
+        if len(self._ndarray) and not lib.is_string_array(self._ndarray, skipna=True):
+            raise ValueError("Must provide strings")
+
+    @classmethod
+    def _from_sequence(cls, scalars, dtype=None, copy=False):
+        if dtype:
+            assert dtype == "string"
+        result = super()._from_sequence(scalars, dtype=object, copy=copy)
+        # convert None to np.nan
+        # TODO: it would be nice to do this in _validate / lib.is_string_array
+        # We are already doing a scan over the values there.
+        result[result.isna()] = np.nan
```

Comment:
Would we consider just failing here if non-string / NA data makes its way through? Not sure our `None` / NA handling is consistent enough to introduce this

Reply:
> Would we consider just failing here if non-string / NA data makes its way through?

It's a bit hidden, but we do fail here.

```pytb
In [4]: pd.arrays.StringArray._from_sequence(['a', 'b', 1])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-26199d15894d> in <module>
----> 1 pd.arrays.StringArray._from_sequence(['a', 'b', 1])

~/sandbox/pandas/pandas/core/arrays/string_.py in _from_sequence(cls, scalars, dtype, copy)
    134         if dtype:
    135             assert dtype == "string"
--> 136         result = super()._from_sequence(scalars, dtype=object, copy=copy)
    137         # convert None to np.nan
    138         # TODO: it would be nice to do this in _validate / lib.is_string_array

~/sandbox/pandas/pandas/core/arrays/numpy_.py in _from_sequence(cls, scalars, dtype, copy)
    149         if copy and result is scalars:
    150             result = result.copy()
--> 151         return cls(result)
    152
    153     @classmethod

~/sandbox/pandas/pandas/core/arrays/string_.py in __init__(self, values, copy)
    123         super().__init__(values, copy=copy)
    124         self._dtype = StringDtype()
--> 125         self._validate()
    126
    127     def _validate(self):

~/sandbox/pandas/pandas/core/arrays/string_.py in _validate(self)
    128         """Validate that we only store NA or strings."""
    129         if len(self._ndarray) and not lib.is_string_array(self._ndarray, skipna=True):
--> 130             raise ValueError("Must provide strings")
    131
    132     @classmethod

ValueError: Must provide strings

```

This is through `StringArray._validate()`, which is called at the end of `StringArray.__init__`.